### ISSUE TITLE
bug report: alias code for group mail aliases

### ISSUE THREAD

**Issue by bibiak1** [docker-mailserver/docker-mailserver#3262]
### Preliminary Checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.
- [X] I searched the issue tracker but was unable to find my issue.
- [X] I read the [extended documentation in general](https://docker-mailserver.github.io/docker-mailserver/latest/) but found nothing to resolve the issue.
- [X] I read the [documentation on debugging](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/), tried the proposed steps to debug the problem, but was still unable to resolve the issue.

### Affected Component(s)

Aliases setup

### What happened and when does this occur?

When I tried to create group mail aliases I came into problem with adding recipient to the group.
Problem appears if there's an email account and alias to redirect mails.

### What did you expect to happen?

```console
root@mail:/# setup email list
* testuser@domain.com ( 0 / ~ ) [0%]

root@mail:/# setup alias list
[  ERROR  ]  '/tmp/docker-mailserver/postfix-virtual.cf' is empty, nothing to list
[  ERROR  ]  Aborting
root@mail:/# setup email add test1@domain.com
Enter Password: 
root@mail:/# setup email add test2@domain.com
Enter Password: 
root@mail:/# setup alias add test1@domain.com test1@other.com
root@mail:/# setup alias add test2@domain.com test2@other.com
root@mail:/# setup alias add group-1@domain.com test1@domain.com
root@mail:/# setup alias add group-1@domain.com test2@domain.com
root@mail:/# setup alias add group-2@domain.com test1@domain.com
root@mail:/# setup alias add group-2@domain.com test2@domain.com
[  ERROR  ]  'group-2@domain.com' is already an alias for recipient: 'test2@domain.com'
[  ERROR  ]  Aborting
root@mail:/# 
```

### How do we replicate the issue?

1. setup email add test1@domain.com
2. setup email add test2@domain.com
3. setup alias add test1@domain.com test1@other.com
4. setup alias add test2@domain.com test2@other.com
5. setup alias add group-1@domain.com test1@domain.com
6. setup alias add group-1@domain.com test2@domain.com
7. setup alias add group-2@domain.com test1@domain.com
8. setup alias add group-2@domain.com test2@domain.com

### DMS version

v12.0.0

### What operating system is DMS running on?

Linux

### Which operating system version?

Docker

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker Compose

### docker-compose.yml

```yml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.example.com
    volumes:
      - /etc/localtime:/etc/localtime:ro
    environment:
      - ENABLE_RSPAMD=1
      - ENABLE_CLAMAV=1
      - ENABLE_FAIL2BAN=1
    cap_add:
      - NET_ADMIN # For Fail2Ban to work
```

### Other relevant information

Solution: https://github.com/docker-mailserver/docker-mailserver/pull/3259

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [X] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by georglauterbach** [comment]
Thanks, now we have a clearer picture! I would like @casperklein and @polarathene to have a look at this as well and get their opinion on the matter :)

**Comment by polarathene** [comment]
**TL;DR:** I think this is not a bug, but a misunderstanding? I am aware of one page in our docs where this kind of example is given and might have been referenced by @bibiak1 ?

If not I'd like a more concrete example of when this isn't a bad idea.

---

https://github.com/docker-mailserver/docker-mailserver/blob/1076aac37d3e30e921f3917b7eb35b461904d1c4/target/bin/addalias#L34-L36

> ```bash
> root@mail:/# setup email add test1@domain.com
> root@mail:/# setup email add test2@domain.com
> root@mail:/# setup alias add test1@domain.com test1@other.com
> root@mail:/# setup alias add test2@domain.com test2@other.com
> ```

Why are you adding mail accounts as aliases? I don't think we're meant to allow for treating a mail account as an alias like that.

You want `group-1` and `group-2` aliases to both receive mail and send it to the `test1` and `test2` aliases or mail accounts? Or you expect to receive on DMS and store for their `domain.com` mail accounts there, and also relay to their `@other.com`?

Alias groups to both test mail accounts, and the external `@other.com`, but don't alias the mail account? Maybe I am mistaken, but it doesn't make sense to me why mail account and alias would overlap like this?

---

I see that we have [this example in docs](https://docker-mailserver.github.io/docker-mailserver/latest/examples/tutorials/basic-installation/#using-dms-as-a-local-mail-relay-for-containers), and perhaps that is incorrect (_it was a community contributed example from years ago_). [Our docs on aliases](https://docker-mailserver.github.io/docker-mailserver/latest/config/user-management/#aliases) does not align with support for a mail account as an alias.

I remember working on the alias refactor and being very thorough on making that correct. I'm curious if it's really a bug, or the failure is correctly preventing it? (_albeit not preventing mail accounts as aliases_)

**Comment by bibiak1** [comment]
Ok, if explanations are needed...

I've my own domain which is supported by your product but I'm forwarding emails to Google for easier management - that's why I have aliases to other.com. But to send emails from from google as my domain I need email account as well - right???

Aliases for groups are to support emails from 3rd party services that suppose to be forwarded to different recipients.

**Comment by georglauterbach** [comment]
No problem :) This is why we want users to raise an issue first 🙂

Our relay code is not perfect, but should work fine for you :)

**Comment by bibiak1** [comment]
Wrong.... my approach was correct as :
- I do not want to send email using Google as relay
- I would like to send emails and use my server as relay with auth - looks like I need local email accounts to achieve that

Summary: my patch is needed

**Comment by polarathene** [comment]
Look at guide like this: https://community.mxroute.com/t/beginners-guide-to-setting-up-mxroute-with-gmail/629

- Gmail will import email sent to DMS for the account, into Gmail inbox. POP3 remove from DMS mail storage after each mail is imported.
- Gmail can be configured to add your `@domain.com` account, and configure login to port 587 or 465 (preferred). This is just like ThunderBird or other mail client. It will send to DMS, with authorization to send mail as your `@domain.com` account.

DNS for your mail should then be the same as if no Gmail involved.

Your group aliases are fine, they will send to the real `@domain.com` accounts. Alias is for that purpose, to deliver to 1 or more accounts (eg: `sales@domain.com` => `john.doe@domain.com`, `jane.doe@domain.com`).

---

<details>
<summary>Original Response</summary>

> I'm forwarding emails to Google for easier management, that's why I have aliases to `@other.com`.

Ok, so you want to manage email from Gmail web/app? You can use POP3, emails go to your DMS and get stored. Gmail then pull them to your gmail account inbox with POP3, removing them from DMS storage. Two separate mail accounts.

There is [guide like this one](https://themeisle.com/blog/use-gmail-with-your-own-domain-name/).

Alternatively, you configure Gmail as your MX host. It will say when mail is sent to `@domain.com`, send to Gmail. Then you need to configure Gmail to understand `@domain.com` belongs to your Gmail account. This would be for all `@domain.com` however.

> But to send emails from from google as my domain I need email account as well - right???

To send mail is different. You use other DNS record to authorize other mail server to send mail as you. Such as SPF records. I think Gmail can support that as well, there should be guide online to lookup.

> Aliases for groups are to support emails from 3rd party services that suppose to be forwarded to different recipients.

Yes, this is what alias is for. Address for non-account, that will be changed to 1 or more mail accounts (local or external).

---

Relay is bit different.

- You send mail from DMS, it can be relayed through another provider like SendGrid or Gmail. Useful when Port 25 is blocked or wanting good sender reputation. Not what you're asking for.
- DMS receive mail from external account like `bibiak1@gmail.com`, two types of relay:
  - **A:** Relay as same `@gmail.com` address because `@gmail.com` has authorized you to send on behalf.
  - **B:** Relay as `@domain.com`. DMS configured to send again as different address, or rewrite address.

I think you want B, but I'm not sure about how to configure that right now. If DMS support that, perhaps someone else can explain.

Alias would not work here for foreign account since DMS is not MX host for `@gmail.com`.

</details>

**Comment by georglauterbach** [comment]
> Please follow procedure to replicate the issue

I won't, because as I already said, we do not support

```console
root@mail:/# setup email add test1@domain.com
Enter Password: 
root@mail:/# setup alias add test1@domain.com test1@other.com
```

No point in going on. But thank you very much for making us aware of this flaw.

> DMS says that I'm not able to add alias because recipient is part of it - which is untruth

This is indeed weird, but not a surprise because we are already in an undefined state when executing the commands above.

---

#3270 will resolve this. Then you can look over this issue again, properly read the feedback from @polarathene and apply it.

**Comment by bibiak1** [comment]
How to setup forwarding after this patch?

**Comment by georglauterbach** [comment]
> How to setup forwarding after this patch?

Can you try <https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1509366419>, and see whether this works for you?

**Comment by bibiak1** [comment]
I do not want gmail to login to the server to get emails

**Comment by georglauterbach** [comment]
> I do not want gmail to login to the server to get emails

Why?

**Comment by diogotcorreia** [comment]
I came across this issue when looking at the changelog for 12.1.0.
I, too, have been using this feature forever -- it was actually the reason I switched to DMS.
My use-case is to have a catch-all address (i.e. `@domain.com` that redirects to an external service).
In most cases, #3270 doesn't break this functionality, but there are some addresses where I have overridden the aliases and they have a matching account (i.e. `user1@domain.com` exists both as an alias and as an email account) in order to be able to send mail as well.
No email is stored on DMS.

An example of this setup would be:
```
setup alias add @domain.com my.account@other.com
setup alias add user1@domain.com user1@other.com
setup account add me@domain.com
setup account add user1@domain.com
```

This has been working flawlessly for the last few years.

I am confused by the suggestions given here, https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1508619730, the second link shows instructions for exactly this, adding both an alias and an email address.
How would we achieve this same functionality with the new update?

I feel like me and @bibiak1 are not the only ones using this feature.

Thank you for all the development throughout the years!

**Comment by polarathene** [comment]
> in order to be able to send mail as well.
> No email is stored on DMS.

As long as `SPOOF_PROTECTION=0` you should be able to send as whatever user you want to, even if no account is made for it. Just have a local account that you can login as to authenticate.

> I am confused by the suggestions given here, [#3262 (comment)](https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1508619730), the second link shows instructions for exactly this, adding both an alias and an email address.

That was a community contributed guide, but was identified as bad advice. Most maintainers are quite busy presently as we only can work on DMS as volunteers, so this has not been revised yet unfortunately.

> My use-case is to have a catch-all address (i.e. `@domain.com` that redirects to an external service).
> I feel like me and @bibiak1 are not the only ones using this feature.

I believe we identified that this was the wrong approach. Lets say you create two accounts for sending mail, while having your catch-all alias domain. Now send mail to the other account, and what will DMS do?
- Match the catch-all and relay it to the external service?
- Recognize a local Dovecot account was created for storing the mail and avoiding the relay?

I forget which has priority, but I think it is Dovecot, which is not what you'd probably expect / want right? (_even if you'd not send mail to another account this way, another user might attempt to and expect it to work_)

---

The alias approach for this is using the feature incorrectly in that sense AFAIK. Especially with how @bibiak1 was configuring it, and the community guide.

The suggested approach I provided to @bibiak1 was to [configure your external mail account (typically Gmail) to pull email from DMS via POP3](https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1509366419). I assume this isn't liked for a catch-all where you allow any address, or when there are many accounts to configure retrieving mail from into a single external account, I can understand that.

---

> How would we achieve this same functionality with the new update?

You could create your accounts and manually add add aliases to `postfix-virtual.cf` to bypass our `setup` tool. If you need to add another account later, remove that alias line and repeat. Alternatively just temporarily remove the accounts or alias config files, should achieve the same outcome.

You could alias specific accounts and have your "admin" account to login DMS and send mail as any email address from the catch-all domain (technically any domain).

```bash
setup alias add @domain.com my.account@other.com
setup alias add user1@domain.com user1@other.com

setup account add me@domain.com
# send as user1 from me@domain.com account
```

**Comment by georglauterbach** [comment]
PS: I hid my initial comment that contains the ill advice. Again, sorry for that.
